rfd/0080-piv-private-key.md

```diff
@@ -0,0 +1,316 @@
+}
+```
+
+### Certificate private key policy extension
+
+When the Auth Server receives a new certificate request, it will check the cluster auth preference and the user's role options to get their required private key policy. If their policy is `hardware_key`, then the Auth Server will look for a valid attestation in the backend associated with the given public key. If their policy is `hardware_key_touch`, then the Auth Server will also check that the attestation has `TouchRequired: true`.
+
+If the attestation checks are successful, then the Auth Server will sign the TLS and SSH certificates with the new `private_key_policy` TLS extra subject name and Extension respectively. The Auth Server's Authorizer, which is used for all API requests, will check certificates for this extension before authorizing requests.
+
+### Client changes
+
+Teleport clients will need the ability to connect to a user's hardware key, generate/retrieve private keys, and use those keys for cryptographical operations. 
+
+#### Private key policy discovery
+
+Teleport clients should be able to automatically determine if a user requires a hardware private key for login to avoid additional UX concerns. For this, we will introduce the `GetPrivateKeyPolicy` rpc.
+
+```proto
+service AuthService {
+  rpc GetPrivateKeyPolicy(GetPrivateKeyPolicyRequest) returns (GetPrivateKeyPolicyResponse);
+}
+
+// The user will be pulled from the request certificate, so request message can be empty for now.
+message GetPrivateKeyPolicyRequest {}
+
+message GetPrivateKeyPolicyResponse {
+    PrivateKeyPolicy policy = 1;
+}
+
+enum PrivateKeyPolicy {
+    PRIVATE_KEY_POLICY_NONE = 0;
+    PRIVATE_KEY_POLICY_HARDWARE_KEY = 1;
+    PRIVATE_KEY_POLICY_HARDWARE_KEY_TOUCH = 2;
+}
+```
+
+In order for a Teleport Client to call `GetPrivateKeyPolicy`, it first needs a set of valid certificates. Unlike `IsMFARequired`, which issues an MFA certificate Reissue request after logging in with non-MFA certificates, we need to call `GetPrivateKeyPolicy` on the initial login.
+
+One option is to make add an endpoint in the Proxy Web API which can be called before login. In order to authenticate the user, we would need them to login with their login credentials. For MFA and Password login, we can potentially reuse these password credentials to intiate the actual login directly after retrieving the user's private key policy requirement. However, for sso login and some edge cases is MFA login (OTP token timing, between requests, etc.), the user would be required to perform the login twice.
+
+Alternatively, we could provide a user with set of certificates using a plain RSA private key in memory first. The resulting certificates will fail the hardware key attestation check in the Auth server, so they will be invalid for all API requests, except for `GetPrivateKeyPolicy` and `GenerateUserCerts`. This way, after the initial login, `GenerateUserCerts` can be used to issue new certificates with a hardware private key.
```

Comment:
This assuming some configuration mismatch, correct? As in, the user doesn't have PIV configured the way Auth wants?

This pattern of "authn -> get insufficient cert -> authn in some other form -> get better cert" is something I'm dealing with for Device Trust as well. In the specific scenario we have here, I wonder if it's better to just error and let `tsh` handle the PIV requirements, so we don't make authn more complex for a one-time setup.

Reply:
> This assuming some configuration mismatch, correct? As in, the user doesn't have PIV configured the way Auth wants?

This will happen on any login that requires `private_key_policy: hardware_key+`, because we have no (secure) way of retrieving the user's required policy before login, and we are not providing a flag/config option for the client to enable PIV login.

We could always try `private_key_policy: hardware_key` if a PIV key is connected, but then we are making `tsh` users with `private_key_policy: none` use PIV anyways, and users with `private_key_policy: touch` would still fail. I think we are better off being naive first and trying normal disk-based key login first, and then having `tsh` handle PIV requirements from there.

> I wonder if it's better to just error and let tsh handle the PIV requirements, so we don't make authn more complex for a one-time setup.

If we want to make the user re-auth, then I agree that giving an error message with the required private key policy would be simpler solution than using the `GetPrivateKeyPolicy` rpc.

> so we don't make authn more complex for a one-time setup.

Whether this is a one-time setup or not is hard to say. We could have `tsh` check the user's current key for the cluster and user, and automatically enable PIV if it was used in the previous login, but other than that this re-auth flow will be undergone for every new login.

I'm still trying to find a way to simplify this flow, so let me know if you have any idea. For now I'll move towards the error and re-auth approach.

Reply:
Happy to chat/brainstorm anytime. Also feel free to ping me on Slack about this.


My initial, probably naive idea, is something like this:


__Authn without PIV (bootstrap)__

1. tsh initiates login flow, collects credentials, but uses a "soft" key

2. Auth emits a certificate without PIV extensions, but realizes that there are PIV restrictions present and tells tsh that in the response

    aka, we relay back the private_key_mode it failed to meet

3. tsh tells the user they are missing PIV requirements, asks for a Yubikey, does bootstrapping/AttestHardwareKey and tries to login again

    We'd get a second password/MFA prompt here.

    tsh records PIV info in its configs.

4. Auth emits a cert with PIV extensions


Note that Web would simply ignore the PIV hints and keep using the non-PIV certificate.


__Authn with PIV (all except bootstrap)__

1. tsh initiates the login flow, either reads Yubikeys and finds a Teleport certificate (signifying PIV was used) or find PIV information in configs (whatever is easier)

2. tsh collects credentials and asks the PIV-backed key to be signed

3. Auth emits a cert with PIV extensions